External signer documentation for staking CLI (#1815)
nucypher/nucypher: docs/source/guides/network_node/staking_guide.rst

Staking
--------

Staking transactions can be broadcast using either a local or remote ethereum node.
By default transaction signing requests are forwarded to the configured ethereum provider;
This is the typical configuration for locally or independently run ethereum nodes.

> In order to use a remote ethereum provider (Alchemy, Infura, Public Remote Node) an external transaction signing client

Reviewer: ```suggestion
In order to use a remote ethereum provider (e.g. Alchemy, Infura, Public Remote Node) an external transaction signing client
```

> (like clef or geth) is needed separate from the broadcasting node.

Reviewer: ```suggestion
(e.g. clef or geth) is needed separate from the broadcasting node.
```

Staking
--------

Using Clef as an external transaction signer
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. important::

    External signing support is an experimental feature and under active development.

Motivation
**********

Instead of running a full ethereum node for both transaction signing and broadcast, an external signer can be specified
via command line independent from the provider. This allows stakers and other network users to use hardware and software

> wallets without needing to run an ethereum node at all.

Author: Give it another look with the added heading description under "Staking".  I want to be careful not to take on the tone of a report of *what changed*.  

For example:
> Currently, a full ethereum node needs to be run to perform on-chain transactions.

Is not really true anymore.

> wallets without needing to run an ethereum node at all.

Reviewer: Sorry that was supposed to be "Previously".

> wallets without needing to run an ethereum node at all.

Reviewer: Fair point - updated suggestion.

> wallets without needing to run an ethereum node at all.

Author: Have another look :-)

Some examples:

- Infura/Alchemy/Etc. for broadcasting with clef signer
- Local geth node for broadcasting with clef signer
- Remote ethereum node for broadcasting with local geth signer

Clef Setup
**********

We'll quickly walk through setup steps below, but additional in-depth documentation on clef can
be found in the source repository here https://github.com/ethereum/go-ethereum/tree/master/cmd/clef

Clef is typically installed alongside geth.  If you already have geth installed on you system you
may already have clef installed.  To check for an existing installation run:

.. code:: bash

    $ clef --version
    Clef version 0.0.0

> Next, initialize Clef with your chosen password to encrypt the master seed:

Author: :heavy_check_mark:

.. code:: bash

    $ clef init
    ...
    The master seed of clef will be locked with a password.
    Please specify a password. Do not forget this password!
    Password:

Running Clef for Goerli
***********************

Clef can use hardware wallets (ledger and trezor) over USB, or geth formatted private keys
by specifying the keystore directory path:

.. code:: bash

    $ clef --keystore <PATH TO GOERLI KEYSTORE> --chainid 5 --advanced

- <PATH TO KEYSTORE> - The path to the directory containing geth-formatted private key files.

>   the default for linux is `~/.ethereum/goerli/keystore`.

Author: :heavy_check_mark:

- Chain ID 5 is specified to ensure clef signs transactions with the network ID of Goerli.

.. code:: bash

    Enter 'ok' to proceed:
    > ok
   ...

    ------- Signer info -------
    * extapi_version : 6.0.0
    * extapi_http : n/a
    * extapi_ipc : /home/user/.clef/clef.ipc
    * intapi_version : 7.0.0

Using clef as a nucypher signer
*******************************

.. code:: bash

    $ nucypher <COMMAND> <ACTION> --signer <CLEF IPC PATH>

Some examples:

.. code:: bash

    # Create a new stakeholder with clef as the default signer
    $ nucypher stake init-stakeholder --signer ~/clef/clef.ipc ...

    # Update an existing configuration with clef as the default signer
    $ nucypher stake config --signer ~/clef/clef.ipc  # Set clef as the default signer

    # Create a new stake using inline signer and provider values
    $ nucypher stake create --signer ~/clef/clef.ipc --provider ~/.ethereum/goerli/geth.ipc

Interacting with clef
*********************

Requests for account management, and signing will be directed at clef, with a 10 second timeout.

> Be alert for user-interactive requests from the clef CLI.

Reviewer: ```suggestion
Be alert for user-interactive requests and confirmations from the clef CLI.
```

Author: :heavy_check_mark:
